aep/general/0122/aep.md

```diff
@@ -0,0 +1,318 @@
+# Resource paths
+
+Most APIs expose _resources_ (their primary nouns) which users are able to
+create, retrieve, and manipulate. Additionally, resources have _paths_: each
+resource has a unique identifying path that users use to reference that
+resource, and these paths are what users should _store_ as the canonical path
+for the resources.
+
+## Guidance
+
+All resource paths defined by an API **must** be unique within that API. (See
+the section on [full resource paths](#full-resource-paths) below for more
+information on referring to resources across APIs.)
+
+Resource paths are formatted according to the [URI path schema][], but without
+the leading slash:
+
+    publishers/123/books/les-miserables
+    users/vhugo1802
+
+- Resource path components **should** usually alternate between collection
```

Comment:
Singletons!

Reply:
Also I generally wonder if singleton's don't belong in the "design patterns" category rather than the "spec" category: do clients do any special handling for singletons today? do they need to?

Reply:
What makes a singleton special is:

* Path differences
  * Different structure
  * Singular rather than plural naming
* Cannot support Create/List/Delete

I'm not sure clients need to handle any of that in a special way, as long as they have the path formats for the singleton and all of its descendants, and _don't_ assume this **should** is a **must**.

Actually...I suppose you might want an SDK that lets you write something like:

```
// Normal sub-resource
bars = foo.listBars();

// Singleton
fooConfig = foo.getConfig();
```

The GAPIC generator would need to know about singletons in order to generate `getConfig`, right?

Reply:
yes, it's true that you would have to know that getConfig is a method. But that could also be achieved by, say, a custom method.

I'm thinking beyond SDKs, which have all the information about the API up-front: there are scenarios where there are lightweight clients: shell scripts or highly dynamic clients which may only have a yaml / json file representing a resource, but would benefit from being able to update it, or list all the other resources associated with it.

Path structure for a singleton doesn't *have* to be different - you could do the repeating structure, and just implement a singleton as a single resource with a static id (e.g. "singleton", "default", or "instance").

Reply:
> yes, it's true that you would have to know that getConfig is a method. But that could also be achieved by, say, a custom method.

Why would we want a custom method rather than the standard `Get` method?

> Path structure for a singleton doesn't have to be different - you could do the repeating structure, and just implement a singleton as a single resource with a static id (e.g. "singleton", "default", or "instance").

This is true...but I think it tests the tradeoff between API aesthetics and client-friendliness.  It's not even that clients _can't_ handle singletons, it's just a bit of extra work.  I might be missing your point about lightweight clients though...if we're assuming a description of a resource, why can't that description say whether the resource is a singleton?

Reply:
for a lightweight client, I'm thinking about use cases where you don't even have the description readily available. Something like a shell script that:

1. reads the list of resources in a particular collection.
2. filters and decides to update a value.
3. modifies the value and performs a subsequent update on each one.

This particular example isn't a singleton, nor does it requires alternating collections, but I'm just noting that shell scripts are an example of a very lightweight client.

I have a few esoteric ones for collection pairs, but if I can think of a simple one I'll reply here with it.

Reply:
Hmm...in that example, the shell script must have some hard-coded knowledge of the structure of the API, right?  It knows the collection URL, it knows about the List and Update operations, and it knows the resource schema.  So...now the knowledge of how the API works (including how singletons work) is in the script author's head rather than in the script.  No?

But I know you pointed out that this isn't an example specific to this question.  Would definitely find it helpful to have one!

Reply:
Discussed live:  change to **must** but add an explicit carve-out for singletons.